packages/apollo-datasource-rest/src/connectors/memcached.ts

```diff
@@ -0,0 +1,79 @@
+import { KeyValueCache } from '../KeyValueCache';
+import Memcached from 'memcached';
+import { promisify } from 'util';
+
+export default class MemcachedKeyValueCache implements KeyValueCache {
+  readonly client;
+  readonly defaultSetOptions = {
+    ttl: 300,
+    tags: [],
+  };
+  // might have to deal with issues of overflow, use BigInt?
+  private logicalClock = 0;
+
+  constructor(serverLocation: Memcached.Location, options?: Memcached.options) {
+    this.client = new Memcached(serverLocation, options);
+    // promisify client calls for convenience
+    this.client.get = promisify(this.client.get).bind(this.client);
+    this.client.getMulti = promisify(this.client.getMulti).bind(this.client);
+    this.client.set = promisify(this.client.set).bind(this.client);
+    this.client.flush = promisify(this.client.flush).bind(this.client);
+  }
+
+  async set(
+    key: string,
+    data: string,
+    options?: { ttl?: number; tags?: string[] },
+  ): Promise<void> {
+    const { ttl, tags } = Object.assign({}, this.defaultSetOptions, options);
+
+    // augment data with tags and version
+    const version = ++this.logicalClock;
+    const payload = {
+      v: version,
```

Comment:
Hmmm, this wasn't really what I had in mind, and I don't think this will work as is. If we keep a logical clock in memory, that will be per process. So it seems there is no way to keep this working correctly across multiple server instances talking to the same store.

Reply:
My idea was that we would store `<tag>: <version>` pairs in cache entries, and also keep separate `<tag>: <version>` entries. We could then use store primitives to safely increment the version for a tag entry without the need for a global clock.

Reply:
I was thinking that it would be easy to move the logical clock into the key-value store as an entry of its own, once we think about multiple server processes. It seems to me like an easier thing to keep track of?

Reply:
If we have `<tag>: <version>` pairs, wouldn't that be overwritten every time there we write an entry with the same tags? In that case, the version number would be different from what was stored with other cache entries.

Reply:
Hmmm, interesting! I'm worried keeping a global version instead of a version per tag will become a bottleneck and/or source of concurrency issues, but other people may have a better idea of the impact. I suspect that will depend on the characteristics of the store. 

On Memcache, one thing to take into consideration is that the global version key will only be stored on a single node, so that will receive a lot of requests. And if it goes down, you basically lose the ability to validate all data (although the same is true for a subset of the data for per-tag version keys).